crates/acvm/src/lib.rs

```diff
@@ -38,178 +40,200 @@ pub trait Backend: SmartContract + ProofSystemCompiler + PartialWitnessGenerator
-                            );
-                            false
+                    _ => GateResolution::Skip,
+                },
+                Directive::Oddrange { a, b, r, bit_size } => match initial_witness.get(a) {
+                    Some(val_a) => {
+                        let int_a = BigUint::from_bytes_be(&val_a.to_bytes());
+                        let pow: BigUint = BigUint::one() << (bit_size - 1);
+                        if int_a >= (&pow << 1) {
+                            return GateResolution::UnsatisfiedConstrain;
                         }
-                        _ => true,
-                    },
+                        let bb = &int_a & &pow;
+                        let int_r = &int_a - &bb;
+                        let int_b = &bb >> (bit_size - 1);
+
+                        initial_witness
+                            .insert(*b, FieldElement::from_be_bytes_reduce(&int_b.to_bytes_be()));
+                        initial_witness
+                            .insert(*r, FieldElement::from_be_bytes_reduce(&int_r.to_bytes_be()));
+                        GateResolution::Resolved
+                    }
+                    _ => GateResolution::Skip,
                 },
+            },
+        }
+    }
+
+    fn solve(
+        &self,
+        initial_witness: &mut BTreeMap<Witness, FieldElement>,
+        mut gates_to_resolve: Vec<Gate>,
+    ) -> GateResolution {
+        let mut unresolved_gates: Vec<Gate> = Vec::new();
+        let mut pass_number = 0;
+        let mut ctx = BinarySolver::new();
+        //binary_solve is used to manage the binary solving mode:
+        //binary_solve.is_none()                => binary solve is not activated
+        //binary_solve == Some(forward_pass)    => binary solve is activated and will do forward_pass forward passes, decreasing at each pass until it reach 0 where the process will be backward
+        let mut binary_solve = None;
+
+        while !gates_to_resolve.is_empty() {
+            unresolved_gates.clear();
+
+            let mut process = |gate| {
+                let mut result = self.solve_gate(initial_witness, gate);
+                if binary_solve.is_some() && result == GateResolution::Skip {
+                    result = ctx.solve(gate, initial_witness);
+                }
+                result
             };
-            if unsolved {
-                unsolved_gates.push(gate);
+
+            let gates: Box<dyn Iterator<Item = _>> = if binary_solve == Some(0) {
+                //we go backward because binary solver should execute only when the program returns an array
+                //in that case it is a bit more efficient to go backwards, although both ways work.
+                Box::new(gates_to_resolve.iter().rev())
+            } else {
+                Box::new(gates_to_resolve.iter())
+            };
+            for gate in gates {
+                match process(gate) {
```

Comment:
Looks like the closure is unnecessary now

Reply:
Done